firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java

```diff
@@ -558,7 +558,13 @@ public void releaseQuery(Query query) {
             queryCache.updateQueryData(queryData);
           }
 
-          localViewReferences.removeReferencesForId(queryData.getTargetId());
+          // We remove both locally edited documents and documents that are associated via the query
+          // cache.
```

Comment:
I would add more to this comment. Something like: "We may have determined that a document is associated with a target, but if that hasn't been confirmed by the backend, the query cache won't know to remove the reference for those documents. Make sure we remove them here."

Or some variation on that.

Reply:
Updated the comment to:

// References for documents sent via Watch are automatically removed when we delete a
// query's target data from the reference delegate. Since this does not remove references
// for locally mutated documents, we have to remove the target associations for these
// documents manually.